Sources/BSON/DocumentIterator.swift

```diff
@@ -21,14 +21,30 @@ public struct DocumentIterator: IteratorProtocol {
     private var buffer: ByteBuffer
 
     internal init(over buffer: ByteBuffer) {
-        fatalError("Unimplemented")
+        self.buffer = buffer
+        _ = self.buffer.readInteger(as: Int32.self) // put the readerIndex at the first key
     }
 
     /// Advances to the next element and returns it, or nil if no next element exists.
     public mutating func next() -> (String, BSON)? {
```

Comment:
I think we need to work on error case handling a bit in this method. we can't throw due to protocol restrictions, but in cases where something goes wrong, IMO we should fatal error with a descriptive message rather than just returning nil.

any document we are iterating over here has either been explicitly validated already (via the public initializer) or is a document we ourselves have chosen to bypass validation for.

WDYT @patrickfreed ?

Reply:
I think that given the contract you mentioned, `fatalError` here seems okay. It just means that we need to be really careful with the places that we bypass validation (which we should have been doing anyways).